test: update the confirm lb script to support private agents

Aashiq-J opened the pull request:
### Description

Issue: https://github.com/terraform-ibm-modules/terraform-ibm-base-ocp-vpc/issues/438

<!--- Replace this text with a summary of the changes in this PR. Include why the changes are needed and context about the changes. List required dependencies. If there is a Git issue for the change, please link to it. --->

### Release required?
<!--- Identify the type of release. For information about the changes in a semantic versioning release, see [Release versioning](https://terraform-ibm-modules.github.io/documentation/#/versioning). --->

- [ ] No release
- [ ] Patch release (`x.x.X`)
- [x] Minor release (`x.X.x`)
- [ ] Major release (`X.x.x`)

##### Release notes content

<!--- If a release is required, replace this text with information that users need to know about the release. Write the release notes to help users understand the changes, and include information about how to update from the previous version.

Your notes help the merger write the commit message for the PR that is published in the release notes for the module. --->

### Run the pipeline

If the CI pipeline doesn't run when you create the PR, the PR requires a user with GitHub collaborators access to run the pipeline.

Run the CI pipeline when the PR is ready for review and you expect tests to pass. Add a comment to the PR with the following text:

```
/run pipeline
```

### Checklist for reviewers

- [ ] If relevant, a test for the change is included or updated with this PR.
- [ ] If relevant, documentation for the change is included or updated with this PR.

### For mergers

- Use a conventional commit message to set the release level. Follow the [guidelines](https://terraform-ibm-modules.github.io/documentation/#/merging.md).
- Include information that users need to know about the PR in the commit message. The commit message becomes part of the GitHub release notes.
- Use the **Squash and merge** option.

Aashiq-J commented:
/run pipeline

Aashiq-J commented:
/run pipeline

Aashiq-J commented:
/run pipeline

vburckhardt commented on scripts/confirm_lb_active.sh:
```diff
@@ -22,28 +25,69 @@ if [[ -z "${RESOURCE_GROUP_ID}" ]]; then
-    sleep 5
-done
+if [ "$PRIVATE_ENV" ]; then
+    until ibmcloud login -q -r "us-south" -a private.cloud.ibm.com || [ $attempts -ge 3 ]; do
```
It should be possible to get the token through  
data "ibm_iam_auth_token" "tokendata" {}

which I think is going through the private IAM token endpoint if the provider is configured so.

vburckhardt commented on main.tf:
```diff
@@ -490,12 +493,29 @@ locals {
+  identifier = var.vpc_id
+}
+
+resource "ibm_is_virtual_endpoint_gateway" "private_is_vpe" {
```
I don't think the module should create the VPE here. The VPEs already exist in the environment in which the schematics agent is running. The whole effort is to go through those existing VPE rather than creating new ones.

vburckhardt commented on main.tf:
```diff
@@ -490,12 +493,29 @@ locals {
 
   provisioner "local-exec" {
-    command     = "${path.module}/scripts/confirm_lb_active.sh ${var.region} ${var.resource_group_id} ${local.lbs_associated_with_cluster[count.index]}"
+    command     = "${path.module}/scripts/confirm_lb_active.sh ${var.region} ${var.resource_group_id} ${local.lbs_associated_with_cluster[count.index]} ${var.private_environment} ${ibm_is_virtual_endpoint_gateway.private_is_vpe[0].service_endpoints[0]}"
```
At minimum the VPE url should be an input of the module. but suggest to investigate if we could actually (1) build that url statically in the module (2) or obtain it.

vburckhardt reviewed (commented):
Thanks for the PR. I added a few comments / guidance.

Aashiq-J commented:
/run pipeline

Aashiq-J commented:
<img width="1714" alt="Screenshot 2024-05-30 at 8 17 01 PM" src="https://github.com/terraform-ibm-modules/terraform-ibm-base-ocp-vpc/assets/122446118/9dd5c462-707d-403a-b765-f87cd9d4657d">

tested in a private vsi with no outbound access

ocofaigh commented on variables.tf:
```diff
@@ -19,6 +19,12 @@ variable "region" {
 
+variable "private_environment" {
+  type        = bool
+  description = "Enable if using private schematics agents."
```
It should probably not mention schematics agents here. Thats not a requirement. Its just the runtime has to have access to private endpoint

ocofaigh commented on variables.tf:
```diff
@@ -19,6 +19,12 @@ variable "region" {
   description = "The IBM Cloud region where the cluster will be provisioned."
 }
 
+variable "private_environment" {
```
should it be called something like `use_private_endpoint` ?

ocofaigh commented on scripts/reset_iks_api_key.sh:
```diff
@@ -22,13 +23,25 @@ if [[ -z "${RESOURCE_GROUP_ID}" ]]; then
-    sleep 5
-done
+if [ "$PRIVATE_ENV" ]; then
+    until ibmcloud login -q -r "us-south" -a private.cloud.ibm.com || [ $attempts -ge 3 ]; do
```
this is missing the resource group scope

ocofaigh commented on main.tf:
```diff
@@ -490,15 +493,19 @@ locals {
     environment = {
-      IBMCLOUD_API_KEY = var.ibmcloud_api_key
+      # IBMCLOUD_API_KEY = var.ibmcloud_api_key
+      IAM_TOKEN = data.ibm_iam_auth_token.tokendata.iam_access_token
```
other option is to add a depends_on on the data lookup so it only runs after cluster is created but just before the script runs

vburckhardt replied:
@ocofaigh I'd rather rely on the provider whenever possible and avoid bash.

ocofaigh replied:
looks like depends_on was used with data lookup which should suffice

Aashiq-J commented:
/run pipeline

vburckhardt commented on modules/fscloud/variables.tf:
```diff
@@ -19,6 +19,12 @@ variable "region" {
 
+variable "use_private_endpoint" {
+  type        = bool
+  description = "Enable if the runtime has to have access to private endpoints."
```
I'd specify exactly what is meant by "runtime".

Suggest to use "Enable if the Terraform execution environment requires access to private endpoints."

Aashiq-J commented:
/run pipeline

Aashiq-J commented:
/run pipeline

Aashiq-J commented:
/run pipeline

Aashiq-J commented:
/run pipeline

Aashiq-J commented:
/run pipeline

Aashiq-J commented:
/run pipeline

ocofaigh commented on scripts/reset_iks_api_key.sh:
```diff
@@ -22,13 +23,25 @@ if [[ -z "${RESOURCE_GROUP_ID}" ]]; then
-    sleep 5
-done
+if [ "$PRIVATE_ENV" ]; then
+    until ibmcloud login -q -r "us-south" -g "${RESOURCE_GROUP_ID}" -a private.cloud.ibm.com || [ $attempts -ge 3 ]; do
```
looks like you hard coded `"us-south"` here, but you used `"${REGION}"` in else statement

Aashiq-J replied:
<img width="831" alt="Screenshot 2024-06-12 at 1 14 01 PM" src="https://github.com/terraform-ibm-modules/terraform-ibm-base-ocp-vpc/assets/122446118/079b4dc0-cb71-4017-9c9b-21a5c9bae101">

ocofaigh replied:
might be worth putting that as a code comment

ocofaigh commented on variables.tf:
```diff
@@ -19,6 +19,12 @@ variable "region" {
 
+variable "use_private_endpoint" {
+  type        = bool
+  description = "Enable if the Terraform execution environment requires access to private endpoints."
```
Can we come up with a better description here? Like what exactly will happen if this is set to true? Something like:

> Set this to true to force all api calls to use the IBM Cloud private endpoints

ocofaigh commented on modules/fscloud/variables.tf:
```diff
@@ -19,6 +19,12 @@ variable "region" {
 
+variable "use_private_endpoint" {
+  type        = bool
+  description = "Enable if the Terraform execution environment requires access to private endpoints."
```
same as https://github.com/terraform-ibm-modules/terraform-ibm-base-ocp-vpc/pull/442/files#r1634669406

vbontempi commented on scripts/confirm_lb_active.sh:
```diff
@@ -3,47 +3,43 @@
-fi
+LB_ID="$2"
+PRIVATE_ENV="$3"
+API_VERSION="2024-03-01"
```
is the version fixed because related to a specific requirement?
otherwise we could set to (today - 1) to avoid to stick to a fixed value that we have to maintain in the future

vbontempi commented on scripts/reset_iks_api_key.sh:
```diff
@@ -22,13 +23,25 @@ if [[ -z "${RESOURCE_GROUP_ID}" ]]; then
-    echo "Error logging in to IBM Cloud CLI..." >&2
-    sleep 5
-done
+if [ "$PRIVATE_ENV" ]; then
```
given that the difference is only the cloud endpoint if private env is true or false, wouldn't be better to set the cloud endpoint value according to the private env value and don't repeat the entire code in if and else statement?

vbontempi reviewed (commented):
left a couple of comments

Aashiq-J commented:
/run pipeline

Aashiq-J commented:
/run pipeline

Aashiq-J commented:
@ocofaigh , @vbontempi 
- Removed the need of ibmcloud_api_key in the root module. Using token to fetch all the keys and run the reset logic.
- Found the private api to reset the api key
- removed the dependency on API_VERSION date.

ocofaigh reviewed (changes requested):
Main readme usage needs to be updated since you removed apikey variable: https://github.com/terraform-ibm-modules/terraform-ibm-base-ocp-vpc/tree/vpe-test?tab=readme-ov-file#usage

ocofaigh commented on main.tf:
```diff
@@ -48,6 +48,8 @@ locals {
   # attach_ibm_managed_security_group is false and custom_security_group_ids is set => only use the custom security group ids
   cluster_security_groups = var.attach_ibm_managed_security_group == true ? (var.custom_security_group_ids == null ? null : concat(["cluster"], var.custom_security_group_ids)) : (var.custom_security_group_ids == null ? null : var.custom_security_group_ids)
+  # tflint-ignore: terraform_unused_declarations
+  validate_private_endpoint = var.use_private_endpoint == true && var.cluster_config_endpoint_type == "default" ? tobool("When setting 'var.use_private_endpoint' to true, 'var.cluster_config_endpoint_type' must be set to private.") : true
```
We can double check with @toddgiguere but I'm not sure "default" = public endpoint. For example, if you have a private only cluster, I think "default" would be private (but I might be wrong)

ocofaigh reviewed (changes requested):
@Aashiq-J see latest comments

ocofaigh reviewed (changes requested):
Also @Aashiq-J It seems we are not testing `use_private_endpoint` being set to true in any example. I suggest you set it to true in the fscloud example, and also maybe update that example to add additional security group IDs so we also test the `confirm_lb_active.sh` running in private only mode

Aashiq-J commented on scripts/confirm_lb_active.sh:
```diff
@@ -3,47 +3,39 @@
+    STATUS=$(curl -H "Authorization: $IAM_TOKEN" -X GET "$URL" | jq -r '.operating_status')
+    echo "Load balancer status: $STATUS"
+    if [[ "$STATUS" == "online" ]]; then
+        sleep 300
```
This sleep is set once we get an online status for the first time.  We have noticed that, even after getting online status and the script passes, the loadbalancer would go onto another update stage before the execution is back to terraform. Hence, added this sleep to check for online status twice with the gap of 5 mins to check if the lb is not going onto another update state.

Aashiq-J commented on scripts/reset_iks_api_key.sh:
```diff
@@ -22,29 +17,63 @@ if [[ -z "${RESOURCE_GROUP_ID}" ]]; then
+    # Update the global apikeys variable
+    apikeys=$combined_data
+
+    echo "$apikeys"
```
This doesn't print anything on the terminal, it is to return the value of the string when we call the fetch function.

Aashiq-J commented:
/run pipeline

Aashiq-J commented:
/run pipeline

vbontempi reviewed (commented):
in my opinion the fetch_data function may be designed in a different way to avoid to return a list of apikeys to be filtered later: maybe it could still retrieve the list of apikeys but it could simply compare each one with APIKEY_KEY_NAME (got as input param) and then it could simply return a different return code if the apikeys must be reset or not.
Moreover this would avoid also the dependency on jq.

This is just an opinion
By looking at the current code I haven't see something to fix, probably the RIAS version currently hardcoded could be calculated.

Aashiq-J commented:
/run pipeline

vbontempi commented:
> 1bd2340

Considering this comment, the script looks ok now

Aashiq-J commented:
/run pipeline

terraform-ibm-modules-ops commented:
:tada: This PR is included in version 3.26.0 :tada:

The release is available on [GitHub release](https://github.com/terraform-ibm-modules/terraform-ibm-base-ocp-vpc/releases/tag/v3.26.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
